### PR TITLE
Improve timing collection and granularity

### PR DESCRIPTION
**Pull request by ausmith**
Today we are collecting API timings, but have instances in our code where we are calling through the additional `lib/cb/client.rb` that exists in this project. This is not useful to us, so we are going to dig a level beyond that file to see who's calling it.

Also move the observer notification to a `begin...ensure` block to avoid losing the timings around errors. This is still valuable information to know and should impact our API timings dashboard numbers.

### CODE REVIEW

File: spec/cb/utils/api_spec.rb [hunk 1]
```diff
@@ -101,10 +139,10 @@ module Utils
             expect(api).to receive(:notify_observers).twice.and_call_original
             expect(Cb::Models::ApiCall).to receive(:new).with(:cb_post_before, '/moom', {},
                                                               { file: __FILE__, method: 'block (4 levels) in <module:Utils>' },
-                                                              nil, 0.0).at_most(1).times.and_call_original
+                                                              nil, 0.0).once.and_call_original
             expect(Cb::Models::ApiCall).to receive(:new).with(:cb_post_after, '/moom', {},
                                                               { file: __FILE__, method: 'block (4 levels) in <module:Utils>' },
-                                                              { success: 'yeah' }, instance_of(Float)).at_most(1).times.and_call_original
+                                                              { success: 'yeah' }, instance_of(Float)).once.and_call_original
```
Comment (tkling) [inline spec/cb/utils/api_spec.rb]: +1, dig these tiny changes to lower the chaining count

File: lib/cb/utils/api.rb [hunk 2]
```diff
@@ -57,10 +57,14 @@ def cb_delete(path, options = {}, &block)
       def execute_http_request(http_method, uri, path, options = {}, &block)
```
Comment (tkling) [inline lib/cb/utils/api.rb]: This method is starting to grow kinda lengthy. I see a possible small change that could happen - breaking into 2 methods, one that starts the timing and sets up stuff like `api_caller` and eventing (at least begin, maybe both), and another method that handles the wrapped execution and returns the response. It's more of a nice to have, style type thing though. I can read through it the way it is now, and the tests make it very clear that the logic here is correct.

Comment (tkling) [inline lib/cb/utils/api.rb]: Just wanted to throw that out there as a thing to ponder and perhaps spend some time on - I wouldn't be heartbroken if it merged the way it is now.

**Comment by tkling** [comment]
Just curious - have you hooked this up to c-d or c-m locally to see observers work appropriately with an API error? The new specs make it pretty clear that things are executing as expected, just wondering.

File: lib/cb/utils/api.rb (hunk 2 above)
Comment (ausmith) [inline lib/cb/utils/api.rb]: I'm not opposed to doing that, but that's larger than this PR. What I'd imagine for breaking this apart would be breaking `exeucute_http_request` to `wrapped_http_request` and `execute_http_request`.
`wrapped_http_request`: Does the timing setup portion and calls out to `execute_http_request`.
`execute_http_request`: Does the request itself and the response validation call.

So I say future optimization there.

**Comment by ausmith** [comment]
I have not done that as the errors we're trying to catch here are the unknown unknowns :stuck_out_tongue:

**Comment by tkling** [comment]
You'd still be able to verify that timing and stuff works ok though, right?

**Comment by tkling** [comment]
It could be as simple as changing a client or response method to always throw an error, then just verifying that c-m still sees timing for that call.

**Comment by ausmith** [comment]
Working on it

File: lib/cb/utils/api.rb (hunk 2 above)
Comment (tkling) [inline lib/cb/utils/api.rb]: Not sure I agree that it's larger than this PR - now is as good a time as ever for cleaning up this method. I would agree if this method wasn't being changed as part of this PR, but it is :) It's not really a time-sensitive PR either, so taking an extra hour or so to clean it up would be good in the long run as we wouldn't have to add a card to revisit it later.

**Comment by ausmith** [comment]
For anyone following along, there's an issue in c-d with how we are trying to source the oauth2 gem (there are 2 of them because of our branch off the upstream version). Going to work on that tomorrow so we can smoothly transition once this guy is merged.

File: lib/cb/utils/api.rb (hunk 2 above)
Comment (tkling) [inline lib/cb/utils/api.rb]: After your explanation last night (everybody and their brother is mocking the `execute_http_request` method) I agree. Let's do this in another PR, it'll be easier to review this in 2 chunks like you're saying :thumbsup:

**Comment by ausmith** [comment]
How I proved this: figured out the correction to our internal c-d project's Gemfile, then used the local path to the gem with a `fail` inside the `begin...ensure` block here and was able to get the observer to print to the rails log. This is ready to roll!

**Comment by tkling** [comment]
:thumbsup: Will merge after CC is done running
